### PR TITLE
Added clipping feature to Arc, Ellipse, Image, Polygon, Rectangle and Sprite

### PR DESCRIPTION
**Pull request by nestorrente**
I deleted last commits from `develop-patch-clipping` then the previous pull request was automatically closed.

My branch is in a good state now. I think you can do the merge! :grinning:

Thanks for teaching me something more about using git :smiley:

### CODE REVIEW

File: src/displayobjects/rectangle.js [hunk 1]
```diff
@@ -16,13 +16,18 @@
...
 				// Do fill if a color is specified
 				if (this.fill !== "") {
 					canvas.fillStyle = this.fill;
-					canvas.fillRect(x, y, this.width, this.height);
+					canvas.fill();
 				}
-				
+
+				// // Save canvas (for clipping after draw the stroke)
+				// canvas.save();
```
Comment (koggdal) [inline src/displayobjects/rectangle.js]: Why is this here and commented out? Will the stroke interfere with the clipping?

**Comment by koggdal** [comment]
Cool! :) Could you also add `clipChildren: false` as the default in all objects that support it? (here: https://github.com/nestorrente/ocanvas/blob/develop-patch-clipping/src/displayobjects/image.js#L21)

**Comment by nestorrente** [comment]
Done! :grinning:

I also removed the "// canvas.save()" line in rectangle.js. At first, I though that the stroke will interfere with the clipping, but not.

Somethig else?

**Comment by koggdal** [comment]
Great! :)

Now you committed the built file too though. I should probably fix the build script a bit to not write the files to that location unless you tell it to make a real build (when releasing a new version). Version 2.7.5 is released, so that file should not be modified now. So just remove that file from the commit and then I can improve the build script a bit.

**Comment by nestorrente** [comment]
I think I don't understand what you mean. Which files I must remove from the commit, exactly? The `dist` directory?

**Comment by koggdal** [comment]
Yes exactly. That's really only used for the final builds of a version and should not be modified after the release has been made. Currently the build script will create the file in the `dist` folder, but I will fix it so that it creates the file in a different folder, and put that in `.gitignore` to prevent committing them.

**Comment by koggdal** [comment]
Okay, I have now fixed the build script (you'll get it if you rebase on top of latest `develop` branch).

This will put files in build/dev/

```
node build/build.js
```
    
This will put files in build/dist/

```
node build/build.js release
```

Which means when you run the regular `node build/build.js` while developing, your built files will be put in `build/dev/` (which is ignored by git).

**Comment by nestorrente** [comment]
I think now is all right. Checkout the last commit.

I think I start to understand the way that Git works :grinning: at the beginning I didn't know how to solve a conflict :sweat_smile:

**Comment by koggdal** [comment]
Okay great that you're learning! :) If you're looking at the commits in this pull request now, you'll see that there are many many commits, including commits by me. So you probably didn't rebase correctly.

**Comment by nestorrente** [comment]
OK, I think I am lost again :sweat_smile:

Can you check the Network Graph to know what I did?

https://github.com/koggdal/ocanvas/network

**Comment by koggdal** [comment]
You most likely didn't rebase on top of my changes, but merged them in or something.

So it's probably easier to step back a bit.. :) It looks like what we want these two commits, right?

"Added clipping feature to display objects Arc, Ellipse, Image, Polygo…"
"Added clipChildren:false as the default value"

Then you can use interactive rebase to remove the other commits and to squash these two commits into one, so that the final Pull Request will contain a single commit.

First, make sure you have a good version of `develop` (assuming the repo at koggdal/ocanvas is named upstream as the remote):

```
git checkout develop
git reset --hard upstream/develop
```

Then use interactive rebase (this will pop you into an editor, just like `git commit` would):

```
git checkout develop-patch-clipping
git rebase develop -i
```

When you're in interactive rebase in your editor, just remove the lines for the commits you don't want anymore. Then you should be left with two commits. Both will say `pick` at the start of the line. Change the bottom one to `f`. That will squash the second one into the first one and use the commit message of the first one. Then save and you should be in a good state with a single new commit.

Then push it to your fork (assuming your fork is named origin as the remote):

```
git push -f origin develop-patch-clipping
```

**Comment by nestorrente** [comment]
I never did a iterative rebase until now. It's great!

This last commit, de453ba, must have all the necessary changes to add the "clipChildren" feature.

**Comment by koggdal** [comment]
Great! :smile: Only one thing left now! When I said that you should remove the dist folder, I didn't mean remove the files completely, but remove the changes of them from your commit :) Your commit should only change a few files, not remove anything. To remove the removal of the files (start from the root folder of the repo):

```
// Reset any changes of the build folder to what it was 1 commit before
git reset HEAD~1 build

// Save that in the same commit
git commit --amend

// Remove any unstaged changes (the reset command will put the removed files here)
git checkout .
```

Then push again with `-f` :)
